openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -23,122 +27,201 @@
 
 
-def capture_call_stack(current_model):
-    """ logs customised call stacks in global dictionary `STACK_BOOK`, and logs it.
+def capture_call_stack(entity_name):
+    """ Logs customised call stacks in global dictionary STACK_BOOK and logs it.
 
-    Args:
-        current_model - Name of the model class
+    Arguments:
+        entity_name - entity
     """
-    # holds temporary callstack
-    # frame[0][6:-1] -> File name along with path
-    # frame[1][6:] -> Line Number
-    # frame[2][3:] -> Context
-    temp_call_stack = [(frame[0][6:-1],
-                        frame[1][6:],
-                        frame[2][3:])
-                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+    # Holds temporary callstack
+    # List with each element 4-tuple(filename, line number, function name, text)
+    # and filtered with respect to regular expressions
+    temp_call_stack = [frame for frame in traceback.extract_stack()
                        if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
 
-    # avoid duplication.
-    if temp_call_stack not in STACK_BOOK[current_model] and TRACK_FLAG \
-            and not issubclass(current_model, tuple(HALT_TRACKING)):
-        STACK_BOOK[current_model].append(temp_call_stack)
-        log.info("logging new call stack for %s:\n %s", current_model, temp_call_stack)
+    final_call_stack = "".join(traceback.format_list(temp_call_stack))
 
+    def _should_get_logged(entity_name):  # pylint: disable=
+        """ Checks if current call stack of current entity should be logged or not.
+
+        Arguments:
+            entity_name - Name of the current entity
+        Returns:
+            True if the current call stack is to logged, False otherwise
+        """
+        is_class_in_halt_tracking = (bool(HALT_TRACKING) and
+                                     inspect.isclass(entity_name) and
+                                     issubclass(entity_name, tuple(HALT_TRACKING[-1])))
+
+        is_function_in_halt_tracking = (bool(HALT_TRACKING) and
+                                        not inspect.isclass(entity_name) and
+                                        any((entity_name.__name__ == x.__name__ and
+                                             entity_name.__module__ == x.__module__) for x in tuple(HALT_TRACKING[-1])))
+
+        is_top_none = bool(HALT_TRACKING) and HALT_TRACKING[-1] is None
+
+        if is_top_none:
+            return False
+
+        if HALT_TRACKING:
+            if is_class_in_halt_tracking or is_function_in_halt_tracking:
+                return False
+            else:
+                if temp_call_stack not in STACK_BOOK[entity_name]:
+                    return True
+                else:
+                    return False
+        else:
+            if temp_call_stack not in STACK_BOOK[entity_name]:
+                return True
+            else:
+                return False
```

Comment:
You can remove the `if HALT_TRACKING` condition. If `HALT_TRACKING` is empty, then both `is_class_in_halt_tracking` and `is_function_in_halt_tracking` will be false, so this whole block boils down to:

`return not (is_class_in_halt_tracking or is_function_in_halt_tracking) and temp_call_stack not in STACK_BOOK[entity_name]`